assertionerror when trying to compute reversed cumulative sum

<!-- Please include a self-contained copy-pastable example that generates the issue if possible.

Please be concise with code posted. See guidelines below on how to provide a good bug report:

- Craft Minimal Bug Reports http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports
- Minimal Complete Verifiable Examples https://stackoverflow.com/help/mcve

Bug reports that follow these guidelines are easier to diagnose, and so are often handled much more quickly.
-->

**Describe the issue**:

**Minimal Complete Verifiable Example**:

```python
import pandas as pd
import dask.dataframe as dd

df = dd.from_pandas(pd.DataFrame({'a': [1,2,3], 'b': [4,5,6]}), npartitions=2)
df['a'][::-1].cumsum()
```
throws
```
In [7]: df['a'][::-1].cumsum().compute()
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
Cell In[7], line 1
----> 1 df['a'][::-1].cumsum().compute()

File ~/polars-api-compat-dev/.venv/lib/python3.12/site-packages/dask/dataframe/dask_expr/_collection.py:4134, in Series.__getitem__(self, key)
   4132 if isinstance(key, Series) or self.npartitions == 1:
   4133     return super().__getitem__(key)
-> 4134 return self.loc[key]

File ~/polars-api-compat-dev/.venv/lib/python3.12/site-packages/dask/dataframe/dask_expr/_indexing.py:84, in LocIndexer.__getitem__(self, key)
     81 if isinstance(cindexer, np.generic):
     82     cindexer = cindexer.item()
---> 84 return self._loc(iindexer, cindexer)

File ~/polars-api-compat-dev/.venv/lib/python3.12/site-packages/dask/dataframe/dask_expr/_indexing.py:103, in LocIndexer._loc(self, iindexer, cindexer)
    100 iindexer = self._maybe_partial_time_string(iindexer, unit=unit)
    102 if isinstance(iindexer, slice):
--> 103     return self._loc_slice(iindexer, cindexer)
    104 elif is_series_like(iindexer) and not is_bool_dtype(iindexer.dtype):
    105     return new_collection(LocList(self.obj, iindexer.values, cindexer))

File ~/polars-api-compat-dev/.venv/lib/python3.12/site-packages/dask/dataframe/dask_expr/_indexing.py:157, in LocIndexer._loc_slice(self, iindexer, cindexer)
    155 def _loc_slice(self, iindexer, cindexer):
    156     assert isinstance(iindexer, slice)
--> 157     assert iindexer.step in (None, 1)
    158     return new_collection(LocSlice(self.obj, iindexer, cindexer))

AssertionError: 
```

**Anything else we need to know?**:

**Environment**:

- Dask version: 2025.2.0
- Python version: 3.12
- Operating System: linux
- Install method (conda, pip, source): pip

--- a/dask/dataframe/dask_expr/_indexing.py
+++ b/dask/dataframe/dask_expr/_indexing.py
@@ -200,7 +200,7 @@ class ReverseDataFrame(Expr):
                 _reverse_partition,
                 TaskRef((self.frame._name, npartitions - i - 1)),
             )
-            for i in range(self.npartitions)  # type: ignore
+            for i in range(self.npartitions)
         }
 
 

--- a/dask/dataframe/dask_expr/_indexing.py
+++ b/dask/dataframe/dask_expr/_indexing.py
@@ -15,10 +15,12 @@ from dask.dataframe.dask_expr import from_dask_array
 from dask.dataframe.dask_expr._collection import Series, new_collection
 from dask.dataframe.dask_expr._expr import (
     Blockwise,
+    Expr,
     MaybeAlignPartitions,
     Partitions,
     Projection,
     are_co_aligned,
+    plain_column_projection,
 )
 from dask.dataframe.dispatch import meta_nonempty
 from dask.dataframe.indexing import (
@@ -154,13 +156,52 @@ class LocIndexer(Indexer):
 
     def _loc_slice(self, iindexer, cindexer):
         assert isinstance(iindexer, slice)
+        if iindexer.step is not None and iindexer.step < 0:
+            obj = ReverseDataFrame(self.obj)
+            iindexer = slice(iindexer.start, iindexer.stop, -iindexer.step)
+        else:
+            obj = self.obj
         assert iindexer.step in (None, 1)
-        return new_collection(LocSlice(self.obj, iindexer, cindexer))
+        return new_collection(LocSlice(obj, iindexer, cindexer))
 
     def _loc_element(self, iindexer, cindexer):
         if iindexer < self.obj.divisions[0] or iindexer > self.obj.divisions[-1]:
             raise KeyError("the label [%s] is not in the index" % str(iindexer))
         return new_collection(LocElement(self.obj, iindexer, cindexer))
+
+
+def _reverse_partition(df):
+    return df.loc[::-1]
+
+
+class ReverseDataFrame(Expr):
+    _parameters = ["frame"]
+
+    @functools.cached_property
+    def _meta(self):
+        return self.frame._meta
+
+    def _divisions(self):
+        return (None,) * (self.frame.npartitions + 1)
+
+    @functools.cached_property
+    def npartitions(self):
+        return self.frame.npartitions
+
+    def _simplify_up(self, parent, dependents):
+        if isinstance(parent, Projection):
+            return plain_column_projection(self, parent, dependents)
+
+    def _layer(self) -> dict:
+        npartitions = self.npartitions
+        return {
+            (self._name, i): Task(
+                TaskRef((self._name, i)),
+                _reverse_partition,
+                TaskRef((self.frame._name, npartitions - i - 1)),
+            )
+            for i in range(self.npartitions)  # type: ignore
+        }
 
 
 class LocBase(Blockwise):
@@ -277,6 +318,8 @@ class LocEmpty(LocList):
 
 
 class LocSlice(LocBase):
+    _projection_passthrough = True
+
     @functools.cached_property
     def start(self):
         if self.iindexer.start is not None:

--- a/dask/dataframe/dask_expr/tests/test_indexing.py
+++ b/dask/dataframe/dask_expr/tests/test_indexing.py
@@ -4,6 +4,7 @@ import numpy as np
 import pytest
 
 from dask.dataframe.dask_expr import from_pandas
+from dask.dataframe.dask_expr.io import FromPandas
 from dask.dataframe.dask_expr.tests._util import _backend_library, assert_eq
 
 pd = _backend_library()
@@ -174,6 +175,17 @@ def test_loc_slicing():
     df = from_pandas(pdf, npartitions=npartitions)
     result = df["2024-03-01":"2024-09-30"]["A"]
     assert_eq(result, pdf["2024-03-01":"2024-09-30"]["A"])
+
+
+def test_reverse_indexing(df, pdf):
+    assert_eq(df.loc[::-1], pdf.loc[::-1])
+    result = df.loc[::-1][["x"]]
+    expected = pdf.loc[::-1][["x"]]
+    assert_eq(result, expected)
+    expr = result.expr.optimize(fuse=False)
+    assert expr.frame.frame.columns == ["x"] and isinstance(
+        expr.frame.frame, FromPandas
+    )
 
 
 def test_indexing_element_index():

--- a/dask/dataframe/dask_expr/tests/test_indexing.py
+++ b/dask/dataframe/dask_expr/tests/test_indexing.py
@@ -186,6 +186,14 @@ def test_reverse_indexing(df, pdf):
     assert expr.frame.frame.columns == ["x"] and isinstance(
         expr.frame.frame, FromPandas
     )
+    with pytest.raises(
+        ValueError, match="Can not use loc on DataFrame without known divisions"
+    ):
+        df.loc[1::-1].loc[5]
+    with pytest.raises(
+        ValueError, match="Can not use loc on DataFrame without known divisions"
+    ):
+        df.loc[:5:-1].loc[5]
 
 
 def test_indexing_element_index():

